### ISSUE TITLE
Vis.Js Timeline Performance

### ISSUE THREAD

**Issue by Unfound** [almende/vis#2177]
Hi,
I have encountered a significant performance decrease when moving to Vis.js from Chaplinks. I have created two identical examples that create 10 timelines with 400 items on each timeline.

If no one can get around to fixing it I can fix it with some direction from you guys. Any help would be greatly appreciated.

Issue Links / Examples:
Production Version Using Chaplinks: http://www.archive.arm.gov/discovery/#v/results/s/fdsc::30ebbr
Stack Overflow: http://stackoverflow.com/questions/40094369/vis-js-timeline-performance
Jsfiddle Vis.js: https://jsfiddle.net/33noajkk/1/
Jsfiddle Chaplinks: https://jsfiddle.net/crve71cd/

Thanks!!!

**Comment by yotamberk** [comment]
Hey @Unfound, The performance issue has been solved in the develop branch. This issue should be fixed on the next release.

**Comment by mojoaxel** [comment]
> Production Version Using Chaplinks: http://www.archive.arm.gov/discovery/#v/results/s/fdsc::30ebbr

Nice! Can we add this to our [showcases](http://visjs.org/showcase/index.html)?

> I have encountered a significant performance decrease when moving to Vis.js from Chaplinks.

Until the v.4.17 is released (hopefully in the next weeks) you could try using an older version of vis.js like v4.7.0.

Closing this as a duplicate of #2080

**Comment by mojoaxel** [comment]
> Production Version Using Chaplinks: http://www.archive.arm.gov/discovery/#v/results/s/fdsc::30ebbr

@Unfound By the way: You should set a *min* value for the timelines. I think it makes no sense to scroll too far in the past here ;-)

**Comment by mojoaxel** [comment]
> Production Version Using Chaplinks: http://www.archive.arm.gov/discovery/#v/results/s/fdsc::30ebbr

@Unfound By the way: You should set a *min* value for the timelines. I think it makes no sense to scroll too far in the past here ;-)

**Comment by Unfound** [comment]
Thanks for the quick responses! Yes it can be used in the showcases. The current production version is still using chaplinks. I used 4.7.0 as suggested with no performance increase in load times. The issue I am seeing is the slow load times not the zooming.

Note the following was testing in Firefox with firebug:
Vis.js v4.7.0: 4.58s (onload: 10.64s)
Chaplinks: 3.21s (onload: 5.49s)

I have created a new jsfiddle using that version: https://jsfiddle.net/t1r1v1vk/

**Comment by mojoaxel** [comment]
If your problem are not with fast moving the timeline please try 4.16.1.
I thing in your case there are just too many complicated charts at one page. You should thing about rendering them one after another.

**Comment by yotamberk** [comment]
I agree with @mojoaxel. Consider using something a mind of lazy loading to ease the loading time of the
 latest version isnt solving the problem

**Comment by Unfound** [comment]
The original jsfiddle was using the 4.16.1 and it still had the issue. Ive tried loading one after another while waiting for the previous to complete but that just resulted in them getting added to the page as each completed. The issue appears to be linear where the more timelines you add the slower it gets. Side by side chaplinks is way faster when loading large amounts of data.

Is there any way to get closer to the performance of chaplinks? Unfortunately if we cannot get in the ballpark of chaplinks we will have to stay with chaplinks.

The slowdown occurs when each timeline gets created after the data has already been retrieved.

**Comment by KyleDumas** [comment]
This issue appears to be worse in Safari compared to Firefox and Chrome. I can provide a url to the development version (Proxied for the outside world) if needed via email. Thanks!

**Comment by adrian-ionescu** [comment]
I'm facing this issue as well.

Vis.js graph loads in 10+ seconds, whereas equivalent Chap-links graph loads in 1 second and Google-Timeline virtually instantly.

Looks like RangeItem.redraw() is called for every element during loading (!?). Chrome developer tools even warn you about this: "_Forced reflow is a likely performance bottleneck._"

![visjs](https://cloud.githubusercontent.com/assets/16205848/20569426/576082fe-b1a0-11e6-8c68-940ae999dde9.png)

https://jsfiddle.net/L0g1zy68/3/  -  Vis.js - loads in 10s
https://jsfiddle.net/nLwxqxL8/3/  - Google Timeline - loads in <1s
http://pastebin.com/SXndxUeg - Chap-Links (sorry, couldn't get it to work on jsfiddle) - loads in 1s

Surprisingly, once the timeline loads, moving and zooming works faster with Vis.js than Chap-links.

**Comment by adrian-ionescu** [comment]
I think I found the culprit: https://github.com/almende/vis/blame/88f14a0d42b8af38e0dcd7b9b7d95eaf0df86d67/lib/timeline/component/item/RangeItem.js#L117-L122

I commented those 4 lines out and now the graph loads instantly.

@josdejong, it seems you anticipated this would be a performance problem. Any idea how to fix it? I have no experience with web development myself, unfortunately.

**Comment by jgorene** [comment]
Hello everyone,
I have been already working for some time with Google Visualization API to generate deadlines (among others) ... but I prefer vis.js now for timeline at least ;)

I encounter the same performance problem fully described by @adrian-ionescu 
So far, unfortunately I have been found no solutions...

As soon as a certain number of groups and items are exceeded (eg : +600 events, 32 groups and +60 nested groups) ... the display time is up to 10 times longer than the Google visualization in the same conditions and even over a predefined short period ... like just over a week for example
(CHAP links not yet tested)

**Comment by tomdcc** [comment]
Another +1, we are seeing the same behaviour with many many forced reflows during initial load, and also general slowness when zooming with many items, until we've zoomed in far enough that there aren't as many items visible.
